src/tds/gssapi.c

```diff
@@ -238,8 +240,12 @@ tds_gss_get_auth(TDSSOCKET * tds)
 	 */
 	gss_buffer_desc send_tok;
 	OM_uint32 maj_stat, min_stat;
+#ifdef __APPLE__
+#  define nt_principal *(gss_OID_desc *) GSS_KRB5_NT_PRINCIPAL_NAME
+#else
```

Comment:
This is doing exactly the opposite of what the next line comment is suggesting we want.

Reply:
Ah, sorry, I accidentally left the rationale out of my commit message, which should have included

> * tds_gss_get_auth: Use central GSS_KRB5_NT_PRINCIPAL_NAME on macOS.
>   It's available there with existing build settings, and avoids
> 
>     ld: warning: alignment (2) of atom '_tds_gss_get_auth_ver100.nt_principal'
>     is too small and may result in unaligned pointers 
>
>   which could otherwise crop up dozens of times.

IIRC, the warning occurs only when targeting x86 CPUs; Apple of course switched to "Silicon" ARM CPUs a couple of years ago, but still retains some support for x86.

Reply:
It looks like a compiler/linker bug to me. Why the toolchain is not able to align it properly? It this a problem with data declared inside functions?

Reply:
No, AFAICT the issue is that the relevant `struct` is underaligned on pre-Silicon macOS architectures because
```C
#if defined(__APPLE__) && (defined(__ppc__) || defined(__ppc64__) || defined(__i386__) || defined(__x86_64__))
#pragma pack(push,2)
#endif
```
is in effect for its definition.  I can force the compiler's hand by specifying something like
```C
	static struct {
		TDS_INT8 pad1;
		TDS_INT	 pad2;
		gss_OID_desc oid;
	} nt_principal =
		{ 0, 0,
		  { 10, (void*) "\x2a\x86\x48\x86\xf7\x12\x01\x02\x02\x01" } };

```
to ensure that the structure starts at a suitable phase (4 mod 8) and then referring to `nt_principal.oid`, but that's ugly and adds overhead, whereas my proposed change simply takes advantage of a stock definition that's available anyway.  (I don't know how it arranges to have appropriate alignment, but it evidently does.)

Reply:
Oh... so the bug is in structure definition. Possibly the change simply avoid the warning as linker does not check external structure alignment. Anyway, I would then add a comment about the specific MacOS issue, something like

```c
#ifdef __APPLE__
	/* some MacOS header defines gss_OID_desc with a wrong byte alignment, use external
	 * library definition. */
#  define nt_principal (*(gss_OID_desc *) GSS_KRB5_NT_PRINCIPAL_NAME)
#else
	/* same as GSS_KRB5_NT_PRINCIPAL_NAME but do not require .so library */
	static gss_OID_desc nt_principal = { 10, (void*) "\x2a\x86\x48\x86\xf7\x12\x01\x02\x02\x01" };
#endif
```